src/Configurator/CopyFromRecipeConfigurator.php

```diff
@@ -11,51 +11,96 @@
 
 namespace Symfony\Flex\Configurator;
 
+use Symfony\Flex\Lock;
 use Symfony\Flex\Recipe;
 
 /**
  * @author Fabien Potencier <fabien@symfony.com>
  */
 class CopyFromRecipeConfigurator extends AbstractConfigurator
 {
-    public function configure(Recipe $recipe, $config, array $options = [])
+    public function configure(Recipe $recipe, $config, Lock $lock, array $options = [])
     {
         $this->write('Setting configuration and copying files');
-        $this->copyFiles($config, $recipe->getFiles(), $this->options->get('root-dir'), $options['force'] ?? false);
+        $options = array_merge($this->options->toArray(), $options);
+
+        $lock->add($recipe->getName(), ['files' => $this->copyFiles($config, $recipe->getFiles(), $options)]);
     }
 
-    public function unconfigure(Recipe $recipe, $config)
+    public function unconfigure(Recipe $recipe, $config, Lock $lock)
     {
         $this->write('Removing configuration and files');
-        $this->removeFiles($config, $recipe->getFiles(), $this->options->get('root-dir'));
+        $this->removeFiles($config, $this->getRemovableFilesFromRecipeAndLock($recipe, $lock), $this->options->get('root-dir'));
     }
 
-    private function copyFiles(array $manifest, array $files, string $to, bool $overwrite = false)
+    private function getRemovableFilesFromRecipeAndLock(Recipe $recipe, Lock $lock): array
     {
+        $lockedFiles = array_unique(
+            array_reduce(
+                array_column($lock->all(), 'files'),
+                function (array $carry, array $package) {
+                    return array_merge($carry, $package);
+                },
+                []
+            )
+        );
+
+        $removableFiles = $recipe->getFiles();
+
+        if (!$lockedFiles) {
+            return $removableFiles;
+        }
+
+        foreach ($removableFiles as $file => $data) {
```

Comment:
What about iterating `$lockedFiles` instead? Since the filename already is the key `$removableFiles` that would just be a constant-time lookup.